ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala

```diff
@@ -0,0 +1,128 @@
+        * TODO              Would be more robust and would simplify this code.
+        * TODO              Seems somewhat difficult to do as we want to affect the code generated by scalapb.
+        */
+      override def close(status: Status, trailers: Metadata): Unit = {
+        if (
+          status.getCode == Status.Code.INTERNAL &&
+          (status.getDescription == null || !BaseError.isSanitizedSecuritySensitiveMessage(
+            status.getDescription
+          ))
+        ) {
+          val recreatedException = status.asRuntimeException(trailers)
+          val selfServiceException = LedgerApiErrors.InternalError
+            .UnexpectedOrUnknownException(t = recreatedException)(
+              new DamlContextualizedErrorLogger(logger, emptyLoggingContext, None)
+            )
+            .asGrpcError
+          // Retrieving status and metadata in the same way as in `io.grpc.stub.ServerCalls.ServerCallStreamObserverImpl.onError`.
+          val newMetadata =
+            Option(Status.trailersFromThrowable(selfServiceException)).getOrElse(new Metadata())
+          val newStatus = Status.fromThrowable(selfServiceException)
+          super.close(newStatus, newMetadata)
+        } else {
+          super.close(status, trailers)
+        }
+      }
+
+    }
+    val listener = next.startCall(forwardingCall, headers)
+    new ErrorListener(
+      delegate = listener,
+      call = call,
+    )
+  }
+
+}
+
+class ErrorListener[ReqT, RespT](delegate: ServerCall.Listener[ReqT], call: ServerCall[ReqT, RespT])
+    extends ForwardingServerCallListener.SimpleForwardingServerCallListener[ReqT](delegate) {
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val emptyLoggingContext = LoggingContext.newLoggingContext(identity)
+
+  /** Handles errors arising outside Futures or Akka streaming.
+    */
+  override def onHalfClose(): Unit = {
+    try {
+      super.onHalfClose()
+    } catch {
+      // For StatusException and StatusRuntimeException:
+      // 1. Assuming `t` was produced by self-service error codes and, thus, deeming the corresponding status and
+      //    trailers do not need security sanitization.
+      // 2. We need to catch it and call `call.close` as otherwise gRPC will close the stream with a Status.UNKNOWN
+      //    (see io.grpc.internal.ServerImpl.JumpToApplicationThreadServerStreamListener.internalClose)
+      case t: StatusException =>
+        call.close(t.getStatus, t.getTrailers)
+      case t: StatusRuntimeException =>
+        call.close(t.getStatus, t.getTrailers)
+      case t: Throwable =>
+        val e = LedgerApiErrors.InternalError
+          .UnexpectedOrUnknownException(t = t)(
+            new DamlContextualizedErrorLogger(logger, emptyLoggingContext, None)
+          )
+          .asGrpcError
+        call.close(e.getStatus, e.getTrailers)
+    }
+
+  }
+
+  // TODO error codes: Do onCancel, onComplete, onReady or onMessage also need special error handling?
```

Comment:
Is this planned to do later or..?

Reply:
The tests I wrote pass without any changes in these methods so likely we don't need to touch them.

I'm reluctant to just now add error handling to them unless we can write tests that can verify its behavior.
Also, we might add a try/catch with just a log statement to be able to monitor what, if any, exceptions are passing through these methods.

As for now I consider these something extra to think more about after this PR is merged.